### PR TITLE
feat: implement BLOBHASH opcode

### LINKED ISSUE #55: Opcode: BLOBHASH

### PR DESCRIPTION
**Pull request by ricomateo**
Closes #55 
This PR implements code generation for the [BLOBHASH](https://www.evm.codes/#49?fork=cancun) opcode.

### CODE REVIEW

File: src/syscall.rs [hunk 1]
```diff
@@ -311,6 +311,21 @@ impl<'c> SyscallContext<'c> {
         let log = Log { data, topics };
         self.inner_context.logs.push(log);
     }
+
+    pub extern "C" fn get_blob_hash_at_index(&mut self, index: &U256, blobhash: &mut U256) {
+        // supposing the index is a usize at most
+        // TODO: check if this is ok
+        let index = index.lo as usize;
+
+        if index < self.env.tx.blob_hashes.len() {
```
Comment (JulianVentura) [inline src/syscall.rs:320]: Here you can remove the len check and call `self.env.tx.blob_hashes.get(index)` instead. That will return you an Option which you can match over.

**Review by JulianVentura — APPROVED** [review]
LGTM

**Comment by ricomateo** [comment]
@MegaRedHand this one is ready for review

File: src/syscall.rs [hunk 2]
```diff
@@ -311,6 +311,23 @@ impl<'c> SyscallContext<'c> {
         let log = Log { data, topics };
         self.inner_context.logs.push(log);
     }
+
+    pub extern "C" fn get_blob_hash_at_index(&mut self, index: &U256, blobhash: &mut U256) {
+        // supposing the index is a usize at most
+        // TODO: check if this is ok
+        let index = index.lo as usize;
+        let aux = self.env.tx.blob_hashes.get(index);
```
Comment (MegaRedHand) [inline src/syscall.rs:319]: We can use `unwrap_or_default` here

File: src/syscall.rs [hunk 3]
```diff
@@ -311,6 +311,23 @@ impl<'c> SyscallContext<'c> {
         let log = Log { data, topics };
         self.inner_context.logs.push(log);
     }
+
+    pub extern "C" fn get_blob_hash_at_index(&mut self, index: &U256, blobhash: &mut U256) {
+        // supposing the index is a usize at most
+        // TODO: check if this is ok
+        let index = index.lo as usize;
```
Comment (MegaRedHand) [inline src/syscall.rs:318]: We should return zero when index is bigger than a `usize`

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

File: src/syscall.rs [hunk 4]
```diff
@@ -451,6 +451,17 @@ impl<'c> SyscallContext<'c> {
         }
     }
 
+    pub extern "C" fn get_blob_hash_at_index(&mut self, index: &U256, blobhash: &mut U256) {
+        *blobhash = usize::try_from(index.lo).map_or(U256::default(), |idx| {
```
Comment (MegaRedHand) [inline src/syscall.rs:455]: We should check `index.hi` is zero.

File: src/syscall.rs [hunk 5]
```diff
@@ -451,6 +451,17 @@ impl<'c> SyscallContext<'c> {
...
+            self.env
+                .tx
+                .blob_hashes
+                .get(idx)
+                .cloned()
+                .map_or(U256::default(), |x| U256::from_be_bytes(x.into()))
+        });
```
Comment (MegaRedHand) [inline src/syscall.rs:462]: This is a bit confusing. Generally, we have a chain of transformations (`map`, `and`, etc.), and at the end a single `unwrap`. In this case:

```suggestion
        *blobhash = usize::try_from(index.lo).and_then(|idx| {
            self.env
                .tx
                .blob_hashes
                .get(idx)
        })
        .map(U256::from_be_bytes)
        .unwrap_or_default();
```

**Review by MegaRedHand — CHANGES_REQUESTED** [review]
Left some comments

File: src/syscall.rs (hunk 5 above)
Comment (JulianVentura) [reply src/syscall.rs:462]: Yep, that's way better. Thanks for the idea.

File: src/env.rs [hunk 6]
```diff
@@ -1,5 +1,5 @@
 use crate::primitives::{Address, Bytes, U256};
-
+type B256 = U256;
```
[the remark this answers is not in the document]
Comment (JulianVentura) [reply src/env.rs:2]: Done

File: src/syscall.rs (hunk 4 above)
Comment (JulianVentura) [reply src/syscall.rs]: Done

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**